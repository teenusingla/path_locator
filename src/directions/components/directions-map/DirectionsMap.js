import React, { Component } from "react";

import { maps } from "../../../common/services";

import "./DirectionsMap.scss";

/**
 * @name DirectionsMap
 * @type {Component}
 * @extends Component React component
 * @description This component display the map and contains the functionality to render the routes
 */
class DirectionsMap extends Component {
  // map container to save the ref to render the map
  mapContainer;
  // containe the map reference after rendering the map
  map;
  // google map api reference
  maps;

  /**
   * @name initMap
   * @description Initialize the map
   */
  initMap = async () => {
    this.maps = await this.props.maps();

    this.map = new this.maps.Map(this.mapContainer, {
      zoom: 11,
      center: { lat: 22.372081, lng: 114.107877 }
    });
  };

  /**
   * @description Prepare Map positions from path points
   * @param path Array of points
   */
  preparePositionsFromPath = path => {
    return path.map(([lat, lng]) => new this.maps.LatLng(lat, lng));
  };

  /**
   * @description Plot the received points on map as route directions
   * @param Object Response object returned from the Api containing the path points
   */
  drawDirections = ({ path }) => {
    const directionsService = new this.maps.DirectionsService();
    const directionsRenderer = new this.maps.DirectionsRenderer();

    directionsRenderer.setMap(this.map);

    const positions = this.preparePositionsFromPath(path);
    const waypoints = positions
      .slice(1, positions.length - 1)
      .map(location => ({ location, stopover: false }));

    // request for the google map directions api
    const request = {
      origin: positions[0],
      destination: positions[positions.length - 1],
      waypoints,
      optimizeWaypoints: true,
      travelMode: this.maps.TravelMode.DRIVING
    };

    // get the route from directionService and then plot with the help of directionsRenderer
    directionsService.route(request, (response, status) => {
      if (status === this.maps.DirectionsStatus.OK) {
        directionsRenderer.setDirections(response);
      } else {
        alert("Error in direction service response");
      }
    });
  };

  componentDidMount() {
    this.initMap();
  }

  getSnapshotBeforeUpdate() {
    const { directions } = this.props;
    if (directions) {
      this.drawDirections(directions);
    }
    return null;
  }

  // do nothing
  componentDidUpdate() {}

  render() {
    return (
      <div className="map-container">
        <div ref={el => (this.mapContainer = el)} />
      </div>
    );
  }
}

DirectionsMap.defaultProps = {
  maps
};

export default DirectionsMap;
