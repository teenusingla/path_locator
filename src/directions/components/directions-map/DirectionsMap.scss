@import '../../../responsive.scss';

.map-container {
    width: calc(100% - 32px);
    padding: 16px;
    height: 500px;

    @include mobile {
        height: 300px;
    }
}

.map-container > div {
    width: 100%;
    height: 100%;
}
