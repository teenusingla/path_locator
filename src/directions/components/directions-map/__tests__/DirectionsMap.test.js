import React from "react";
import renderer from "react-test-renderer";
import DirectionsMap from "../DirectionsMap";

describe("<DirectionsMap /> component", () => {
  it("should match the snapshot", () => {
    const tree = renderer.create(<DirectionsMap />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
