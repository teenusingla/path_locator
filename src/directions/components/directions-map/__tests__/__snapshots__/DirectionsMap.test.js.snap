// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<DirectionsMap /> component should match the snapshot 1`] = `
<div
  className="map-container"
>
  <div />
</div>
`;
